File: src/main/java/com/bazbatlabs/bricksmash/models/Paddle.java
package com.bazbatlabs.bricksmash.models;

public final class Paddle implements Collidable {

    private static final float FRICTION = 0.1f;
    private static final float ACCELERATION = 0.75f;
    private static final float WIDTH = 50.0f;
    private static final float HEIGHT = 10.0f;
    private static final float MAX_SPEED = 5.0f;

    private final Rect bounds;

    private final Vec2 size;
    private final WorldEvents events;

    private Vec2 pos;
    private Vec2 vel;

    private boolean acceleratingLeft;
    private boolean acceleratingRight;

    public Paddle(Rect bounds, float startY, WorldEvents events) {

        this.events = events;

        this.acceleratingLeft = false;
        this.acceleratingRight = false;

        this.bounds = bounds;

        this.size = new Vec2(WIDTH, HEIGHT);

        float startX =
            bounds.left() + (bounds.right() - bounds.left() - this.size.x) / 2f;

        this.pos = new Vec2(startX, startY);

        this.vel = Vec2.ZERO;
    }

    public Vec2 vel() { return vel; }
    public Vec2 pos() { return pos; }
    public Vec2 size() { return size; }
    public Rect bounds() { return new Rect(pos, size); }

    public void update() {

        float accelX = 0.0f;
        if (acceleratingLeft) { accelX -= ACCELERATION; }
        if (acceleratingRight) { accelX += ACCELERATION; }

        Vec2 accel = new Vec2(accelX, 0.0f);

        vel = vel.add(accel).scale(1f - FRICTION);

        if (vel.x > MAX_SPEED) { vel = new Vec2(MAX_SPEED, 0.0f); }
        if (vel.x < -MAX_SPEED) { vel = new Vec2(-MAX_SPEED, 0.0f); }

        pos = pos.add(vel);

        if (pos.x + size.x > bounds.right()) {
            vel = Vec2.ZERO;
            pos = new Vec2(bounds.right() - size.x, pos.y);
        }

        if (pos.x < bounds.left()) {
            vel = Vec2.ZERO;
            pos = new Vec2(bounds.left(), pos.y);
        }
    }

    public void startAccelerating(Direction direction) {
        toggleAcceleration(direction, true);
    }

    public void stopAccelerating(Direction direction) {
        toggleAcceleration(direction, false);
    }

    public Vec2 center() {
        return new Vec2(pos.x + (size.x / 2.0f), pos.y + size.y);
    }

    public void hit() {
        events.enqueue(WorldEvents.Event.PADDLE_HIT);
    }

    public Vec2 deflect(Vec2 vel, Vec2 collision, Axis axis) {
        switch (axis) {
        case X:
            return new Vec2(-vel.x, vel.y);

        case Y:
            float rotation = 0.5f - (collision.x - pos.x) / size.x;
            return (new Vec2(vel.x, -vel.y)).rotate(rotation);

        default:
            return Vec2.ZERO;
        }
    }

    private void toggleAcceleration(Direction direction, boolean start) {
        switch (direction) {
        case LEFT: acceleratingLeft = start; break;
        case RIGHT: acceleratingRight = start; break;
        default: break;
        }
    }

    public boolean isActive() { return true; }
}

File: src/main/scala/models/Paddle.scala
package com.bazbatlabs.bricksmash.models

import com.bazbatlabs.bricksmash.lib._

class Paddle(val fieldBounds: Rect, startY: Float, val events: WorldEvents) extends Collidable {

  val size = Vec2(Paddle.Width, Paddle.Height)

  private var acceleratingLeft = false
  private var acceleratingRight = false

  val startX = fieldBounds.left + (fieldBounds.right - fieldBounds.left - size.x) / 2f

  var pos = Vec2(startX, startY)
  var vel = Vec2.Zero

  def bounds = Rect(pos, size)

  def update() {
    var accelX = 0f

    if (acceleratingLeft) { accelX -= Paddle.Acceleration }
    if (acceleratingRight) { accelX += Paddle.Acceleration }

    val accel = Vec2(accelX, 0f)

    vel = vel.add(accel).scale(1f - Paddle.Friction)

    if (vel.x > Paddle.MaxSpeed) { vel = Vec2(Paddle.MaxSpeed, 0f) }
    if (vel.x < -Paddle.MaxSpeed) { vel = new Vec2(-Paddle.MaxSpeed, 0f) }

    pos = pos.add(vel)

    if (pos.x + size.x > fieldBounds.right) {
      vel = Vec2.Zero
      pos = Vec2(fieldBounds.right - size.x, pos.y)
    }

    if (pos.x < fieldBounds.left) {
      vel = Vec2.Zero
      pos = Vec2(fieldBounds.left, pos.y)
    }
  }

  def startAccelerating(direction: Direction.Value) {
    toggleAcceleration(direction, true)
  }

  def stopAccelerating(direction: Direction.Value) {
    toggleAcceleration(direction, false)
  }

  def center = Vec2(pos.x + (size.x / 2f), pos.y + size.y)

  def hit {
    events.enqueue(Event.PaddleHit)
  }

  def deflect(vel: Vec2, collision: Vec2, axis: Axis.Value) =
    axis match {
      case Axis.X => Vec2(-vel.x, vel.y)
      case Axis.Y => {
        val rotation = 0.5f - (collision.x - pos.x) / size.x
        Vec2(vel.x, -vel.y).rotate(rotation)
      }
      case _ => Vec2.Zero
    }

  def toggleAcceleration(direction: Direction.Value, start: Boolean) {
    direction match {
      case Direction.Left => acceleratingLeft = start
      case Direction.Right => acceleratingRight = start
    }
  }

  def isActive = true
}

object Paddle {
    val Friction = 0.2f
    val Acceleration = 0.65f
    val Width = 26f
    val Height = 8f
    val MaxSpeed = 4f
}
